### PR TITLE
feat: make `Storyblok` a "Generic Class" for typed client instances.

### PR DESCRIPTION
**Pull request by ryami333**
Currently, the getter methods (eg. `getStory`, `getStories` etc) all resolve a `ISbStory`-typed object which does not contain any Space-specific type-information. So it meets the following limitations:

```ts
import StoryblokClient from "storyblok-js-client";

export const storyblokClient = new StoryblokClient({
  accessToken: "MY_TOKEN",
});

storyblokClient.getStory("xxxxxxxxxx", {}).then((response) => {
  const story = response.data.story;

  story.content.component; /* Type: string | undefined */
  response.data.story.content.name; /* "Type: any" */
});
```

The problem is that all of the properties of `response.data.story.content` are ambiguously-typed, so we'd have to type-cast it to make it useful:

```ts
type CatStory = ISbStoryData<ISbComponentType<"cat"> & { name: string }>;
type DogStory = ISbStoryData<ISbComponentType<"dog"> & { name: string }>;
// … etc

type StoryUnion = CatStory | DogStory /* | etc | etc */;

storyblokClient.getStory("xxxxxxxxxx", {}).then((response) => {
  const story = response.data.story as StoryUnion; // <- Type-casting

  story.content.component; // Type: "cat" | "dog"
  story.content.name; // Type: string
});
```

In this pull-request, I make the `Storyblok` a [Generic Class](https://www.typescriptlang.org/docs/handbook/2/generics.html#generic-classes) so that it can be instantiated with this type-information (ideally automatically generated by the Storyblok CLI's [`generate-typescript-typedefs`](https://github.com/storyblok/storyblok-cli?tab=readme-ov-file#generate-typescript-typedefs) command) to have a fully-typed API client object:

```ts
import { CatStory, DogStory {/* … etc … */} } from '@/generated/storyblok';

type StoryUnion = CatStory | DogStory /* | etc | etc */;

export const storyblokClient = new StoryblokClient<StoryUnion>({
  accessToken: "MY_TOKEN",
});

storyblokClient.getStory("xxxxxxxxxx", {}).then((response) => {
  const story = response.data.story; // Type: ISbStoryData<StoryUnion>

  story.content.component; /* Type: "cat" | "dog" */
  response.data.story.content.name; /* "Type: string" */
});
```

Note 1: The generic parameter is optional. If you omit it then the current behaviour is retained. In other words **this is a non-breaking change, and would not require a major release**.

Note 2: _Ideally_ the `StoryUnion` would be automatically generated by `generate-typescript-typedefs` - I might consider opening a separate PR for that over in that repo :)

## Pull request type

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Other (please describe):

## How to test this PR

See the last code-block above. Create an instance of a `Storyblok` object using the generic parameter, passing a type that represents your stories.

## What is the new behavior?

All described above.

### Prior Art

This "typed SDK" pattern is the same one observed by many competitors' respective JS clients, for example:

- [Supabase](https://supabase.com/docs/reference/javascript/typescript-support#using-typescript-type-definitions)
- [Directus](https://docs.directus.io/guides/sdk/getting-started.html#creating-a-composable-client)

### CODE REVIEW

File: src/index.ts
Comment (ryami333) [inline src/index.ts:58]: Not actually deleted - just moved to the `interfaces.ts` file alongside other such interfaces.

File: src/index.ts
Comment (ryami333) [inline src/index.ts:60]: Note that providing a default value like `= ISbContentType` makes this a non-breaking change, because that was the default used previously.

File: src/index.ts
Comment (ryami333) [inline src/index.ts:9]: Note that there is some noise in the diffs here from formatting changes (from formatting the file with Prettier). Once https://github.com/storyblok/storyblok-js-client/pull/831 is merged I can rebase to make the diffs quieter.

Comment (ryami333) [reply src/index.ts:9]: In the meantime - you should "Hide whitespace" in the Github view settings above.

Comment (alvarosabu) [reply src/index.ts:9]: Hey, @ryami333 I actually added a suggestion in your previous PR so we can get rid of that noisy whitespace indentation. Once its merged its gonna make this more easy to review

Comment (ryami333) [reply src/index.ts:9]: Yes, I will rebase this PR as soon as that one is merged.

Comment (ryami333) [reply src/index.ts]: @alvarosabu rebased onto `main`, ready for review.

Comment (alvarosabu) [reply src/index.ts]: Hi @ryami333 thanks for taking care of it. One thing about rebasing. We had problems in the past with our automatic semantic release not grabbing the correct versions due to the use of rebase. In general, use merge commits instead. This ensures that the history remains linear and clear, allowing the semantic release tool to accurately determine the correct versions based on the commit messages.

**Review by alvarosabu — CHANGES_REQUESTED** [review]
Hi @ryami333 thanks a lot for introducing this change as non-breaking and for your time putting together the PR, really appreciate it.

We were discussing it internally, I think having generics will make things a lot easier for users to type responses correctly in conjunction with the type generator.

Normally I would aim for a Generic on the Class level, but considering the class methods would fetch different types or schemas or data, why don't we assert types with generics at method level?

So instead of having to type the response also with a generic like this:

```ts
const client = new StoryblokClient<Foo>();
const story: ISBStoryData<Foo> = client.getStory("/foo",{...})
```

We could do this:

```ts
const storyIDontKnowTheTypeOfInAdvance = client.getStory<Foo>(slug, {...})
```

What do you think? That way, each call will have its own expected type

**Comment by alvarosabu** [comment]
Hi @ryami333 just checking if you had the chance to go through the comments I left. Let me know if you have any questions.

**Comment by ryami333** [comment]
Hi @alvarosabu, sorry I've been away sick for the last couple weeks. Here's my response.

The class-level generic-typing was a very deliberate choice, for the following reasons:

## 1. Method-level generic typing is not _really_ any different to just type-casting

In other words, a call-level API really only enables you to do something that you can already trivially do:

```ts
// Method-level generic-typing enables this:
const fooStory = await client.getStory("/foo", {}).then(result => result.data.story as Foo);
// … which is not functionally different to this (type-casting):
const fooStory = await client.getStory<Foo>("/foo", {}).then(result => result.data.story);
```

Put another way: method-level generic typing is not actually _correct_, because with method-level generic-typing there's nothing protecting developers from getting a story of type `Bar` and accidentally typing it `Foo`. Some developers don't care about this, but those developers can already type-cast, so method-level generic-typing is of no particular benefit to them, and they are therefore not the target audience of this PR.

The target audience of this PR are functional programming advocates who avoid using type-casting at all costs, and prefer to leverage Typescript's [control flow analysis](https://www.typescriptlang.org/docs/handbook/2/narrowing.html), like this:

```ts
const story = await client.getStory("/foo", {}).then(result => result.data.story);

story; // ISbStoryData<StoryUnion>
if (story.component !== "Foo") {
  throw new Error(`Expected story to be of type "Foo", got ${story.component} instead.`);
};

story; // ISbStoryData<Foo> <- Typescript automatically narrowed this union-type!
```

## 2. Method-level generic-typing would actually need _two_ generic params, not just one.

Because `getStory` (etc) returns an `IsbStory<Content>`-shaped object, and the type of the `response.data.story` property is only one piece of the puzzle. The type of `response.data.links` and `response.data.rels` are equally important, and you'd need to parametrize those too, eg:

```ts
import { Foo, Bar, Baz /* etc */ } from "../generated-types";

type StoryUnion = Foo | Bar | Baz */ | etc | etc */
const data = await client.getStory<Foo, Story>("/foo", {}).then(result => result.data);

data.links; // (ISbStoryData<StoryUnion> | ISbLinkURLObject)[]
data.rels; // ISbStoryData<StoryUnion>[]
data.story; // ISbStoryData<Foo>
```

If you have to pass the same `StoryUnion` argument to every single one of these calls it makes more sense just to pass this once during the instantiation of the _client_.

---

In conclusion, I wouldn't advocate for method-level typing because this does not reduce the burden on me as maintainer of a Storyblok project. As a maintainer I want to create a single-point-of-failure for correct Storyblok typing in our applications (the instantiation of the client) instead of `n` possible points-of-failure (one for every `.getStory` call). As long as 1) our types/schema are synced, 2) type-checking passes and 3) there is no type-casting (all of which I can enforce through CI scripts) then I have a high degree of confidence in the type-correctness of my colleagues' (and my own) incoming pull-requests.

One _final_ little point. I  _should_ mention that these two approaches are not actually mutually exclusive. We could reasonably implement both patterns (set the "fallback" type on the class-level, and an override on the method-level) but again: it's not a pattern that I advocate for, so I'm not particularly motivated to make such changes to the pull-request, though I'd happily consult/review if you would make a follow-up PR.

**Comment by ryami333** [comment]
@alvarosabu can we have an update please?

**Comment by alvarosabu** [comment]
> @alvarosabu can we have an update, please?

Hi @ryami333 we are still discussing this internally and fixing the CI/CDs, I will ping you when we have something to comment

**Comment by alvarosabu** [comment]
Hi, @ryami333  so after discussing it with the team and some of our users we decided that we are going to evaluate both approaches towards the upcoming v7 release. This version will include a brand-new internal setup and refactoring (yet maintaining the API, so no big changes coming) which will be the ideal time to consider typing improvements. The goal for the new version is to make it easier to contribute to an up-to-date setup  (+ plus a new cool feature, stay tuned 😉 )

Right now, we're focused on a few crucial priorities and we can't work on this at the moment, but as soon as we get to it (likely beginning Q4) I'll let you know. Would you be interested in revisiting these ideas with us when we're drafting the PR? Thanks a lot for the contribution!

**Pull request closed without merging**
